### PR TITLE
[IMP] account, *: Use dynamic placeholder on account.move.form

### PR DESCRIPTION
**Pull request by Megaaaaaa**
*: account_payments, l10n_ar, l10n_cl, l10n_in_edi, l10n_in_edi_ewaybill, l10n_it_edi, l10n_latam_invoice_document, payment_asiapay

---

Description of the issue this commit addresses:

When there is no move in the current sequence, the moves you create take a slot in the sequence while being drafts, this means that if you do not confirm moves in the order they were created, the second move to be confirmed has a lower sequence number than the first which breaks continuity.

---

Desired behavior after this commit is merged:

This commit makes it so that no sequence number is ever assigned while the move stays in draft. A dynamic placeholder shows the user what the name will be with the sequence "template" but it is only used once the move is confirmed.

For subsequent moves, meaning when a first sequence number has been consumed, the old behavior of showing "Draft" as a name stands but there is still no sequence number consumed.

---

Enterprise PR: https://github.com/odoo/enterprise/pull/73837
task-4241510

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_move.py [hunk 1]
```diff
@@ -889,9 +893,19 @@ def _compute_name(self):
...
 
+    @api.depends('journal_id', 'sequence_number', 'sequence_prefix')
+    def _compute_name_placeholder(self):
+        for record in self:
+            record.name_placeholder_list = "/"
+            if record._get_last_sequence() or record.posted_before or record.state != "draft":
+                record.name_placeholder_form = False
+                record._compute_name()
+            else:
+                record.name_placeholder_form = record._get_starting_sequence()[:-1] + "1"
```
Comment (william-andre) [inline addons/account/models/account_move.py:907]: This is not correct if there is a `sequence_override_regex` set on the journal because there could be a suffix after the sequence number

File: addons/account/models/account_move.py [hunk 2]
```diff
@@ -2268,7 +2282,7 @@ def _onchange_name_warning(self):
     @api.onchange('journal_id')
     def _onchange_journal_id(self):
         if not self.quick_edit_mode:
-            self.name = '/'
+            self.name = '/' if self._get_last_sequence() or self.posted_before else False
```
Comment (william-andre) [inline addons/account/models/account_move.py:2285]: What is the difference between `'/'` and `False`?

Since we can have placeholders everywhere does it still make sense to have `/`  in the database? I believe it was only a technical solution to avoid having a empty cell, which was weird for accountants.
But removing this tech artifact could simplify quite a few things here and there.

File: addons/account/models/account_move.py [hunk 3]
```diff
@@ -889,9 +893,19 @@ def _compute_name(self):
...
+        self.filtered(lambda m: not m.name and not move.quick_edit_mode and move._get_last_sequence()).name = '/'
         self._inverse_name()
 
+    @api.depends('journal_id', 'sequence_number', 'sequence_prefix')
+    def _compute_name_placeholder(self):
+        for record in self:
+            record.name_placeholder_list = "/"
+            if record._get_last_sequence() or record.posted_before or record.state != "draft":
+                record.name_placeholder_form = False
+                record._compute_name()
```
Comment (william-andre) [inline addons/account/models/account_move.py:905]: Never call a compute function explicitly.
Even less so a compute on a stored field from the compute of a non stored field.

File: addons/account/tests/test_sequence_mixin.py [hunk 4]
```diff
@@ -56,23 +56,23 @@ def test_sequence_change_date(self):
         """Change the sequence when we change the date iff it has never been posted."""
         # Check setup
         self.assertEqual(self.test_move.state, 'draft')
-        self.assertEqual(self.test_move.name, 'MISC/15-16/01/0001')
+        self.assertMoveName(self.test_move, 'MISC/15-16/01/0001')
```
Comment (william-andre) [inline addons/account/tests/test_sequence_mixin.py:59]: Note sure why you don't simply check the placeholder?

This test is not really relevant anymore; since we never set the sequence in draft anymore, we should never have to change it once it's posted. (from the docstring of this test)

File: addons/account/tests/common.py [hunk 5]
```diff
@@ -682,6 +682,13 @@ def assertNodeDictEqual(node_dict, expected_node_dict):
             self._turn_node_as_dict_hierarchy(expected_xml_tree),
         )
 
+    def assertMoveName(self, move, expected_name):
```
Comment (william-andre) [inline addons/account/tests/common.py:685]: Not sure if we need this. But if we need it, shouldn't it be only in `addons/account/tests/test_sequence_mixin.py`?

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py [hunk 6]
```diff
@@ -521,7 +521,7 @@ def _export_invoice_vals(self, invoice):
         sales_order_id = 'sale_line_ids' in invoice.invoice_line_ids._fields \
                          and ",".join(invoice.invoice_line_ids.sale_line_ids.order_id.mapped('name'))
         # OrderReference/ID (order_reference) is mandatory inside the OrderReference node !
-        order_reference = invoice.ref or invoice.name
+        order_reference = invoice.ref or invoice.name or invoice.name_placeholder_form
```
Comment (william-andre) [inline addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:524]: Why wouldn't we have a name? Why would it be draft?
Should we really give a name if it's not the right one?


____
*Probably not relevant:
With the previous mechanism, we would have been more confident that the number would stick because the next invoice would be with sequence 2, so no conflict possible (whereas with `name_placeholder_form` you have duplicates when in draft)*

File: addons/account_payment/models/payment_transaction.py [hunk 7]
```diff
@@ -88,7 +88,7 @@ def _compute_reference_prefix(self, provider_code, separator, **values):
             invoice_ids = self._fields['invoice_ids'].convert_to_cache(command_list, self)
             invoices = self.env['account.move'].browse(invoice_ids).exists()
             if len(invoices) == len(invoice_ids):  # All ids are valid
-                prefix = separator.join(invoices.mapped('name'))
+                prefix = separator.join([inv.name or inv.name_placeholder_form for inv in invoices])
```
Comment (william-andre) [inline addons/account_payment/models/payment_transaction.py:91]: Doesn't seem right, it means that you could have had multiple `/` before?

File: addons/l10n_latam_invoice_document/models/account_move.py [hunk 8]
```diff
@@ -106,7 +106,7 @@ def _is_manual_document_number(self):
     def _compute_l10n_latam_document_number(self):
         recs_with_name = self.filtered(lambda x: x.name != '/')
         for rec in recs_with_name:
-            name = rec.name
+            name = rec.name or rec.name_placeholder_form
```
Comment (william-andre) [inline addons/l10n_latam_invoice_document/models/account_move.py:109]: You're supposed to have a name in `recs_with_name` :wink:

File: addons/account/models/account_move.py [hunk 9]
```diff
@@ -866,21 +872,19 @@ def _compute_suitable_journal_ids(self):
...
     def _compute_name(self):
         self = self.sorted(lambda m: (m.date, m.ref or '', m._origin.id))
-
         for move in self:
             if move.state == 'cancel':
                 continue
 
             move_has_name = move.name and move.name != '/'
-            if move_has_name or move.state != 'posted':
+            if move_has_name or move.state != 'posted' and move._get_last_sequence():
```
[the remark this answers is not in the document]
Comment (Megaaaaaa) [reply addons/account/models/account_move.py]: Do you mean the method is not required anywhere anymore or just not here ?

In both cases, don't I still need something like it to detect whether this is the first move of the sequence or not ? I'm not aware of similar methods/fields/tools.

File: addons/account/tests/test_sequence_mixin.py (hunk 4 above)
Comment (Megaaaaaa) [reply addons/account/tests/test_sequence_mixin.py]: I interpreted the test more as a way to check that the sequence respects the dates rather than a way to check that the name is updated on a date change.

I think it's best not to erase this test entirely but i agree the placeholder can be tester instead of the name.

File: addons/account/models/account_move.py (hunk 2 above)
Comment (Megaaaaaa) [reply addons/account/models/account_move.py:2275]: As we wanted to keep the behavior in the form view displaying "Draft" as the name when there is already a move in the sequence, i use False for the field to be empty and able to display a placeholder and '/' as the value that is detected in the template as the sign that "Draft" should be shown instead of the empty field with the placeholder (so that the name cannot be modified once the sequence is started).

Would you rather have '/' be set as the placeholder value to be detected instead of the name value ? and so never have the name '/' anymore ?

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py (hunk 6 above)
Comment (Megaaaaaa) [reply addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:524]: For the name: as i said in another comment, when there is no posted move in the sequence, all names are False so the placeholder is shown. Once a move gets posted, all names should be '/' so "Draft" is displayed in the form view. Depending on whether we change that, it will be adapted

For the draft: I'm not aware of all the systems i'm adapting so i try to change the flows as little as possible but i'm often wondering whether we should not limit specific flows to posted moves. Here, i adapt the `order_reference` value to make sure it is set to something when the move is in draft but seeing this is a method to export an invoice, it could be also solved by just making sure the move is posted (and therefore has a name). I just thought if it was made possible it was probably for a reason and shouldn't change...

And so all in all, i have no choice but to give it a potential wrong order_ref if we keep the possibility to use that method on 'draft' moves. Wyt ?

File: addons/account_payment/models/payment_transaction.py (hunk 7 above)
Comment (Megaaaaaa) [reply addons/account_payment/models/payment_transaction.py:91]: Well... yes ?

**Review by Megaaaaaa — COMMENTED** [review]
i'm not done with your review yet and definitely not done with the PR at all anyway but i thought we could already discuss some things to make the rest faster

File: addons/account/models/account_move.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (william-andre) [reply addons/account/models/account_move.py]: Just here

> don't I still need something like it to detect whether this is the first move of the sequence or not ? 

Why would you? Maybe I'm forgetting something?
We needed to detect if it was the first because the first was set even in draft; but it's not the case anymore.

File: addons/account_payment/models/payment_transaction.py (hunk 7 above)
Comment (william-andre) [reply addons/account_payment/models/payment_transaction.py:91]: This "reference prefix" is used to generate a "unique reference for the transaction" in `_compute_reference`.

Pretty sure it's not supposed to change over time. (I'm not an expert in `payment` though)

File: addons/payment_asiapay/tests/test_payment_transaction.py [hunk 10]
```diff
@@ -27,7 +27,17 @@ def test_reference_is_computed_based_on_document_name(self):
         """ Test the computation of reference prefixes based on the provided invoice. """
         self._skip_if_account_payment_is_not_installed()
 
-        invoice = self.env['account.move'].create({})
+        invoice = self.env['account.move'].create({
+            'move_type': 'entry',
+            'date': '2011-11-02',
+            'line_ids': [
+                (0, None, {
```
Comment (william-andre) [inline addons/payment_asiapay/tests/test_payment_transaction.py:34]:
```suggestion
                Command.create({
```

File: addons/account/models/account_move.py [hunk 11]
```diff
@@ -80,7 +80,7 @@ class AccountMove(models.Model):
     _name = "account.move"
     _inherit = ['portal.mixin', 'mail.thread.main.attachment', 'mail.activity.mixin', 'sequence.mixin', 'product.catalog.mixin']
     _description = "Journal Entry"
-    _order = 'date desc, name desc, invoice_date desc, id desc'
+    _order = 'date desc, name desc, invoice_date desc, id'
```
Comment (william-andre) [inline addons/account/models/account_move.py:83]: Why does the order change here? The new order seems counter intuitive since all the other order clauses are to have the most recent first.

File: addons/account/models/account_move.py [hunk 12]
```diff
@@ -871,27 +873,32 @@ def _compute_name(self):
...
                 move._set_next_sequence()
 
-        self.filtered(lambda m: not m.name and not move.quick_edit_mode).name = '/'
         self._inverse_name()
 
+    @api.depends('date', 'journal_id', 'move_type', 'name', 'posted_before', 'sequence_number', 'sequence_prefix', 'state')
+    def _compute_name_placeholder_form(self):
+        self.name_placeholder_form = False
+        for record in self.filtered(lambda move: not move.name or not move.posted_before):
+            if not (record._get_last_sequence() or record.state != "draft"):
```
Comment (william-andre) [inline addons/account/models/account_move.py:898]: Why are you filtering in 2 different places?
```suggestion
        for record in self.filtered(lambda move: 
            (not move.name or not move.posted_before)
            and not move._get_last_sequence() 
            and move.state == "draft"
        ):
```

```suggestion
        for move in self: 
            if (
                (not move.name or not move.posted_before)
                and not move._get_last_sequence() 
                and move.state == "draft"
            ):
```

Comment (william-andre) [reply addons/account/models/account_move.py:898]: If I'm reading your condition right, the `state` has to be `draft`. So `posted_before` cannot be `True`

File: addons/account/models/account_move.py [hunk 13]
```diff
@@ -871,27 +873,32 @@ def _compute_name(self):
...
-                        continue
-                else:
-                    if move_has_name and move.posted_before or not move_has_name and move._get_last_sequence():
-                        # The move either
-                        # - has a name and was posted before, or
-                        # - doesn't have a name, but is not the first in the period
-                        # so we don't recompute the name
-                        continue
-            if move.date and (not move_has_name or not move._sequence_matches_date()):
+            move_has_name = move.name and move.name != "/"
```
Comment (william-andre) [inline addons/account/models/account_move.py:876]: I feel like there is a lot of useless diff here (changing double/single quotes, moving around `if` blocks whose order shouldn't matter)
It makes the diff harder to read and understand. (and all the issues linked to potential conflicts)

Am I missing something?

File: addons/account/models/account_move.py [hunk 14]
```diff
@@ -116,6 +116,8 @@ def _sequence_year_range_monthly_regex(self):
         tracking=True,
         index='trigram',
     )
+    name_placeholder_form = fields.Char(compute='_compute_name_placeholder_form')
+    name_placeholder_list = fields.Char(default='/')
```
[the remark this answers is not in the document]
Comment (Megaaaaaa) [reply addons/account/models/account_move.py]: Stored is a mistake, but the field is required because the static `placeholder="..."` is not usable in list views. Only way to have a placeholder in a list view is to use the dynamic placeholder combinaison `widget="char_with_placeholder_field" options="{'placeholder_field': 'field'}"` which is kinda overkill since it is used to display static data but yeah, it's required...

File: addons/account/views/account_move_views.xml [hunk 15]
```diff
@@ -578,7 +579,7 @@
                 <xpath expr="//field[@name='name']" position="after">
                     <!-- For draft moves, name only displays '/'. Since there may be moves of differenty types,
                          it is better to use display_name indicating what move it refers to when in draft-->
-                    <field name="display_name" decoration-bf="1" string="Number" context="{'name_with_move_type': True}"/>
+                    <field name="display_name" decoration-bf="1" string="Number" context="{'name_with_move_type': True}" widget="char_with_placeholder_field" options="{'placeholder_field': 'name_placeholder_list'}"/>
```
[the remark this answers is not in the document]
Comment (Megaaaaaa) [reply addons/account/views/account_move_views.xml:582]: `placeholder="..."` isn't available in list views

File: addons/l10n_latam_invoice_document/views/account_move_view.xml [hunk 16]
```diff
@@ -49,12 +49,12 @@
...
                     invisible="(not l10n_latam_use_documents or not l10n_latam_manual_document_number) and (not l10n_latam_use_documents or highest_name or state != 'draft')"
                     readonly="posted_before and state != 'draft'"
-                    required="partner_id and l10n_latam_use_documents and (l10n_latam_manual_document_number or not highest_name)"/>
+                    required="partner_id and l10n_latam_use_documents and (l10n_latam_manual_document_number or not highest_name) and state == 'posted'"/>
             </xpath>
 
             <!-- on latam_documents we use document_number to set name -->
             <field name="name" position="attributes">
-                <attribute name="invisible">name == '/' and not posted_before and not quick_edit_mode</attribute>
+                <attribute name="invisible">not ((state != 'draft') or name_placeholder_form or (name and (posted_before or quick_edit_mode)))</attribute>
```
[the remark this answers is not in the document]
Comment (Megaaaaaa) [reply addons/l10n_latam_invoice_document/views/account_move_view.xml]: In my understanding, we don't need a change on the invisible attribute anymore so i'll just remove the line.

File: addons/account/models/account_move.py (hunk 11 above)
Comment (Megaaaaaa) [reply addons/account/models/account_move.py:83]: If a move neither has a `date`, nor a `name`, nor an `invoice_date`, the document has nothing that makes it the most recent. I think when that's the case, we should keep them in the order of creation so that some matching process that create draft entries then match them keep a relevant order. Not matching first created with second source and second created with first source for example.

File: addons/account/models/account_move.py (hunk 12 above)
Comment (Megaaaaaa) [reply addons/account/models/account_move.py]: Reseting to draft a posted move exactly does that, no ? state == "draft" and "posted_before" == True ?

File: addons/account/models/account_move.py [hunk 17]
```diff
@@ -872,26 +874,34 @@ def _compute_name(self):
...
             move_has_name = move.name and move.name != '/'
-            if move_has_name or move.state != 'posted':
-                if not move.posted_before and not move._sequence_matches_date():
-                    if move._get_last_sequence():
-                        # The name does not match the date and the move is not the first in the period:
-                        # Reset to draft
-                        move.name = False
-                        continue
+            if move_has_name or move.state == 'draft':
+                if move_has_name and not move.posted_before and not move._sequence_matches_date():
```
Comment (william-andre) [inline addons/account/models/account_move.py:878]: When do you enter here, in practice?
When do you have a name if it wasn't posted before?

Or is it to support edge cases in stable?

File: addons/account/models/account_move.py [hunk 18]
```diff
@@ -872,26 +874,34 @@ def _compute_name(self):
...
+                    # The name does not match the date and the move is not the first in the period:
+                    # Reset to draft
+                    move.name = False
+                    continue
                 else:
-                    if move_has_name and move.posted_before or not move_has_name and move._get_last_sequence():
+                    if move_has_name and move.posted_before or move.state == 'draft':
```
Comment (william-andre) [inline addons/account/models/account_move.py:884]: When do you not enter this in practice?

File: addons/l10n_ar/tests/test_manual.py [hunk 19]
```diff
@@ -115,17 +115,6 @@ def test_15_liquido_producto_sales(self):
...
-        # 60, 61, 27, 28, 45, 46
-        # In this case manual numbering should be True and the latam document numer should be required
-        with self.assertRaisesRegex(AssertionError, 'l10n_latam_document_number is a required field'):
-            with Form(self.env['account.move'].with_context(default_move_type='out_invoice')) as invoice_form:
-                invoice_form.ref = "demo_liquido_producto_1: Vendor bill liquido producto (DOC 186)"
-                invoice_form.partner_id = self.res_partner_adhoc
-                invoice_form.invoice_payment_term_id = payment_term_id
-                invoice_form.journal_id = self.journal
-                invoice_form.l10n_latam_document_type_id = doc_27_lu_a
-            invoice = invoice_form.save()
```
Comment (william-andre) [inline addons/l10n_ar/tests/test_manual.py:127]: Why is this removed?

File: addons/l10n_latam_invoice_document/views/account_move_view.xml [hunk 20]
```diff
@@ -49,12 +49,11 @@
                 <field name="l10n_latam_document_number"
                     invisible="(not l10n_latam_use_documents or not l10n_latam_manual_document_number) and (not l10n_latam_use_documents or highest_name or state != 'draft')"
                     readonly="posted_before and state != 'draft'"
-                    required="partner_id and l10n_latam_use_documents and (l10n_latam_manual_document_number or not highest_name)"/>
+                    required="partner_id and l10n_latam_use_documents and (l10n_latam_manual_document_number or not highest_name) and state == 'posted'"/>
```
Comment (william-andre) [inline addons/l10n_latam_invoice_document/views/account_move_view.xml:52]: It is required before it is posted AFAIK

File: addons/account/views/account_move_views.xml [hunk 21]
```diff
@@ -946,9 +947,9 @@
                             <div class="text-warning" invisible="not show_name_warning">The current highest number is <field class="oe_inline" name="highest_name"/>. You might want to put a higher number here.</div>
 
                             <h1>
-                                <field name="name" invisible="name == '/' and not posted_before and not quick_edit_mode" readonly="state != 'draft'" placeholder="Draft"/>
+                                <field name="name" invisible="not ((state != 'draft') or name_placeholder_form or (name and (posted_before or quick_edit_mode)))" readonly="state != 'draft'" options="{'placeholder_field': 'name_placeholder_form'}"/>
```
Comment (william-andre) [inline addons/account/views/account_move_views.xml:950]: This looks complicated.
Since `placeholder_name` should only be computed when a placeholder is needed, I think this should be enough.

```suggestion
                                <field name="name" invisible="not (name or placeholder_name or quick_edit_mode)" readonly="state != 'draft'" options="{'placeholder_field': 'name_placeholder_form'}"/>
```

File: addons/account/views/account_move_views.xml [hunk 22]
```diff
@@ -946,9 +947,9 @@
...
 
-                                <span invisible="state != 'draft' or name != '/' or posted_before or quick_edit_mode">Draft</span>
+                                <span invisible="(state != 'draft') or name_placeholder_form or (name and (posted_before or quick_edit_mode))">Draft</span>
```
Comment (william-andre) [inline addons/account/views/account_move_views.xml:952]:
```suggestion
                                <span invisible="name or placeholder_name or quick_edit_mode">Draft</span>
```

File: addons/account/models/account_move.py (hunk 17 above)
Comment (Megaaaaaa) [reply addons/account/models/account_move.py:877]: On the first move of the sequence, the user is able to enter a manual name to override the sequence template. If after that he changes the date of the invoice, it's possible his override doesn't matches the dates anymore. Then we reset the name so he sees the placeholder again.

File: addons/account/models/account_move.py [hunk 23]
```diff
@@ -872,26 +873,24 @@ def _compute_name(self):
...
-                        # - doesn't have a name, but is not the first in the period
-                        # so we don't recompute the name
-                        continue
-            if move.date and (not move_has_name or not move._sequence_matches_date()):
+            if move_has_name or move.state == 'draft':
+                if move_has_name and not move.posted_before and not move._sequence_matches_date():
+                    # The name does not match the date and the move is not the first in the period:
+                    # Reset to draft
+                    move.name = False
+                    continue
```
Comment (william-andre) [inline addons/account/models/account_move.py:881]:
```suggestion
            if move_has_name and not move.posted_before and not move._sequence_matches_date():
                # The name does not match the date and the move is not the first in the period:
                # Reset to draft
                move.name = False
                continue
```

File: addons/account/models/account_move.py [hunk 24]
```diff
@@ -872,26 +873,24 @@ def _compute_name(self):
...
-        self.filtered(lambda m: not m.name and not move.quick_edit_mode).name = '/'
         self._inverse_name()
 
+    @api.depends('date', 'journal_id', 'move_type', 'name', 'posted_before', 'sequence_number', 'sequence_prefix', 'state')
+    def _compute_name_placeholder_form(self):
+        self.name_placeholder_form = False
+        for record in self.filtered(lambda move: (not move.name or not move.posted_before) and (not move._get_last_sequence() or move.state == "posted")):
+            sequence_format_string, sequence_format_values = record._get_sequence_format_param(record._get_starting_sequence())
+            sequence_format_values['seq'] = sequence_format_values['seq'] + 1
+            record.name_placeholder_form = sequence_format_string.format(**sequence_format_values)
```
Comment (william-andre) [inline addons/account/models/account_move.py:893]: I didn't understand your previous condition, did I make a mistake here?
```suggestion
        for move in self:
            if (
                not move.name
                and not move.posted_before
                and not move._get_last_sequence()
            ):
                sequence_format_string, sequence_format_values = move._get_sequence_format_param(move._get_starting_sequence())
                sequence_format_values['seq'] = sequence_format_values['seq'] + 1
                move.name_placeholder_form = sequence_format_string.format(**sequence_format_values)
            else:
                movename_placeholder_form = False
```

File: addons/l10n_ar/tests/test_manual.py [hunk 25]
```diff
@@ -146,17 +146,6 @@ def test_16_liquido_producto_purchase(self):
...
-
-        with self.assertRaisesRegex(AssertionError, 'l10n_latam_document_number is a required field'):
-            with Form(self.env['account.move'].with_context(default_move_type='in_invoice')) as bill_form:
-                bill_form.ref = "demo_liquido_producto_1: Vendor bill liquido producto (DOC 186)"
-                bill_form.partner_id = self.res_partner_adhoc
-                bill_form.invoice_payment_term_id = payment_term_id
-                bill_form.invoice_date = '2023-02-09'
-                bill_form.journal_id = purchase_not_pos_journal
-                bill_form.l10n_latam_document_type_id = doc_60_lp_a
-            bill = bill_form.save()
```
Comment (william-andre) [inline addons/l10n_ar/tests/test_manual.py:158]: Why is this removed?

File: addons/account/models/account_move.py [hunk 26]
```diff
@@ -116,6 +116,7 @@ def _sequence_year_range_monthly_regex(self):
         tracking=True,
         index='trigram',
     )
+    name_placeholder_form = fields.Char(compute='_compute_name_placeholder_form')
```
Comment (william-andre) [inline addons/account/models/account_move.py:119]:
```suggestion
    name_placeholder = fields.Char(compute='_compute_name_placeholder')
```

File: addons/l10n_latam_invoice_document/models/account_move.py
Comment (william-andre) [inline addons/l10n_latam_invoice_document/models/account_move.py:1]: I think that you also don't want a placeholder here, right?
Something similar to `_compute_made_sequence_gap` 
```python
    def _compute_name_placeholder_form(self):
        use_documents_moves = self.filtered(lambda m: m.journal_id.l10n_latam_use_documents)
        use_documents_moves.name_placeholder_form = False
        if other_moves := self - use_documents_moves:
            super(AccountMove, other_moves)._compute_name_placeholder_form()
```

**Review by william-andre — COMMENTED** [review]
@jjscarafia 
Hello, we are making a change (in stable) regarding how the sequencing of journal entries works. It shouldn't change anything for you, but I'd like to make sure.

Basically, the only relevant difference for you is that **we don't set and store a name on the first move of the journal anymore**. I see there is some logic based on `highest_name` in the views, but I'm not sure if we can just remove it like done here (see the diff in `addons/l10n_latam_invoice_document/views/account_move_view.xml` and `addons/l10n_cl/views/account_move_view.xml`)

File: addons/account/models/account_move.py [hunk 27]
```diff
@@ -872,26 +873,25 @@ def _compute_name(self):
...
+            if move.date and not move_has_name and move.state != 'draft':
                 move._set_next_sequence()
 
-        self.filtered(lambda m: not m.name and not move.quick_edit_mode).name = '/'
         self._inverse_name()
 
+    @api.depends('date', 'journal_id', 'move_type', 'name', 'posted_before', 'sequence_number', 'sequence_prefix', 'state')
+    def _compute_name_placeholder_form(self):
+        for move in self:
+            if (not move.name or not move.posted_before) and not move._get_last_sequence():
```
[the remark this answers is not in the document]
Comment (Megaaaaaa) [reply addons/account/models/account_move.py:889]: The best example as to why it should be `or` and not `and` is if you post the first move of the sequence and then reset it to draft. In that case the placeholder has to be back because the sequence is not fixed anymore but the move has a name so, with the condition using `and`, it will have no placeholder and therefore if the user removes the name. Having no name and no placeholder, the field will be replaced by the span "Draft" making it impossible to edit the sequence template.

But it won't show a placeholder for all draft moves like you say since as from the second move, `not move._get_last_sequence()` is falsy so the `if` is not entered, no placeholder is computed and "Draft" is displayed instead.

File: addons/l10n_latam_invoice_document/views/account_move_view.xml [hunk 28]
```diff
@@ -49,12 +49,11 @@
                 <field name="l10n_latam_document_number"
                     invisible="(not l10n_latam_use_documents or not l10n_latam_manual_document_number) and (not l10n_latam_use_documents or highest_name or state != 'draft')"
                     readonly="posted_before and state != 'draft'"
-                    required="partner_id and l10n_latam_use_documents and (l10n_latam_manual_document_number or not highest_name)"/>
+                    required="partner_id and l10n_latam_use_documents and l10n_latam_manual_document_number"/>
```
[the remark this answers is not in the document]
Comment (Megaaaaaa) [reply addons/l10n_latam_invoice_document/views/account_move_view.xml:52]: `highest_name` used to be a way to get the first move **created** as on the first move of the sequence, whatever  the rest, `highest_name` will always be falsy. As this PR aims to make it so no difference is made in the behavior between the first created move and the others but well on the first ever **posted** move and the others, using ``highest_name` in the required field makes no sense anymore.

File: addons/account/views/account_move_views.xml (hunk 15 above)
[the remark this answers is not in the document]
Comment (william-andre) [reply addons/account/views/account_move_views.xml:581]: Looks like you reverted this change?

File: addons/account/models/account_move.py [hunk 29]
```diff
@@ -872,26 +873,25 @@ def _compute_name(self):
...
-                        continue
-                else:
-                    if move_has_name and move.posted_before or not move_has_name and move._get_last_sequence():
-                        # The move either
-                        # - has a name and was posted before, or
-                        # - doesn't have a name, but is not the first in the period
-                        # so we don't recompute the name
-                        continue
-            if move.date and (not move_has_name or not move._sequence_matches_date()):
+            if move.state == 'draft' and not move.posted_before and not move._sequence_matches_date():
```
Comment (william-andre) [inline addons/account/models/account_move.py:876]:
```suggestion
            if not move.posted_before and not move._sequence_matches_date():
```

File: addons/l10n_it_edi/tests/test_edi_import.py [hunk 30]
```diff
@@ -3,7 +3,7 @@
 
 import uuid
 from freezegun import freeze_time
-from unittest.mock import patch
+from unittest.mock import MagicMock, patch
```
Comment (william-andre) [inline addons/l10n_it_edi/tests/test_edi_import.py:6]: Useless import

File: addons/l10n_it_edi/tests/test_edi_import.py [hunk 31]
```diff
@@ -131,12 +129,18 @@ def test_receive_bill_sequence(self):
         def mock_commit(self):
             pass
 
-        invoices = self.env['account.move'].with_company(self.company).search([('name', '=', 'BILL/2019/01/0001')])
-        self.assertEqual(len(invoices), 0)
+        super_create = self.env.registry['account.move'].create
+        create_moves = []
```
Comment (william-andre) [inline addons/l10n_it_edi/tests/test_edi_import.py:133]:
```suggestion
        created_moves = []
```

File: addons/l10n_it_edi/tests/test_edi_import.py [hunk 32]
```diff
@@ -131,12 +129,18 @@ def test_receive_bill_sequence(self):
...
+
+        def mock_create(self, vals_list):
+            ret = super_create(self, vals_list)
```
Comment (william-andre) [inline addons/l10n_it_edi/tests/test_edi_import.py:136]:
```suggestion
            moves = super_create(self, vals_list)
```

File: addons/account/tests/test_sequence_mixin.py [hunk 33]
```diff
@@ -234,75 +243,61 @@ def test_sequence_draft_first_of_period(self):
...
         new_journal = self.test_move.journal_id.copy()
         new_journal.code = "MISC2"
         copy2.journal_id = new_journal
-        self.assertEqual(copy2.name, 'MISC2/15-16/01/0001')
-
-        with Form(copy2) as move_form:  # It is editable in the form
-            with self.assertLogs('odoo.tests.form') as cm:
-                move_form.name = 'MyMISC/2016/0001'
-            self.assertTrue(cm.output[0].startswith('WARNING:odoo.tests.form.onchange:'))
-            self.assertIn('The sequence will restart at 1 at the start of every year', cm.output[0])
```
Comment (william-andre) [inline addons/account/tests/test_sequence_mixin.py:268]: Why are you removing this?

Comment (Megaaaaaa) [reply addons/account/tests/test_sequence_mixin.py:268]: Well these assertions were made on logs but the message of the logs was never shown to the user. With the new modifications, the message isn't even logged in the case of the test so either that or a different test imo. Here's why in details: 

Before this change, all moves would have a name so in `_onchange_name_warning`, `self._origin.name` would have a value resulting in entering the section of the method that decides which error message is to be show. However, as the move is the first of the journal, it's `highest_name` is falsy and `self.show_name_warning` will always be falsy as a consequence. This means that warning log exists but the message is not shown.

Now, as no name is ever assigned before posting the move, `self._origin.name` is falsy before first posting meaning that the message will still not be shown but it will also not even be logged.

If we don't want to remove this form test, I believe it would require to post the move, create a second one and test the behavior on that one.

Comment (william-andre) [reply addons/account/tests/test_sequence_mixin.py:268]: > were made on logs but the message of the logs was never shown to the user.

That's only in the form emulator for the tests. In practice, a warning is shown to the user.

> Now, as no name is ever assigned before posting the move

Then post it, reset to draft and change the value.
Or create a second one like suggested.

**Review by william-andre — APPROVED** [review]
@robodoo r+

**Comment by william-andre** [comment]
@robodoo r+

**Comment by jjscarafia** [comment]
Thanks for the ping @william-andre !
We'll check deeper but I believe it won't affect anything
